Исаев Сергей

Luvr681 opened the pull request:
Решение на предпроверку

@LevShisterov

Luvr681 committed:
Решение на предпроверку

Luvr681 committed:
Решение на проверку

Luvr681 committed:
Решение на проверку

LevShisterov commented on TagCloudContainer/TagCloudContainer/Forms/TagCloudForm.cs:
```diff
@@ -0,0 +1,51 @@
+    private Graphics _graphics;
+    private MainFormConfig _mainFormConfig; 
+    
+    public TagCloudForm(MainFormConfig mainFormConfig)
```
`IMainFormConfig` должно быть на входе.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Forms/Settings.cs:
```diff
@@ -0,0 +1,40 @@
+    {
+        AddConfigValues();
+        _mainFormConfig.Random = Random;
+        TagCloudForm tagCloudForm = new TagCloudForm(_mainFormConfig);
```
Вот тут, вместо `new TagCloudForm` надо сконфигурировать контейнер DI со всеми зависимостями и получить `TagCloudForm` из него.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Providers/TagCloudProvider.cs:
```diff
@@ -0,0 +1,30 @@
+    public TagCloudProvider(string fileName, Point center, Size standartSize, IMainFormConfig mainFormConfig)
+    {
+        IsValidArguments(fileName, center, standartSize);
+        _tagConfig = new TagConfig(center, standartSize, mainFormConfig);
```
`TagConfig` надо получить как зависимость. (простое правило: никаких `new` внутри конструкторов).

LevShisterov commented on TagCloudContainer/TagCloudContainer/Utils/WordsReader.cs:
```diff
@@ -0,0 +1,56 @@
+public class WordsReader
+{
+    private readonly Dictionary<string, Word> _words = new Dictionary<string, Word>();
+    private IWordConfig _wordConfig = new WordConfig();
```
Здесь тоже, `new` в инициализаторе - должно передаваться как зависимость.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/TagConfig.cs:
```diff
@@ -0,0 +1,114 @@
+        word.Size = TextRenderer
+            .MeasureText(word.Value, new Font(_mainFormConfig.FontFamily, word.Weight * _standartSize.Width));
+        
+        if (word.Size.IsEmpty || word.Size == null)
+            throw new ArgumentException("Word size can't be empty or null");
+        if (word.Size.Width < 0 || word.Size.Height < 0) 
+            throw new ArgumentException("Word size can't have negative side value");
```
`word.Size` тут не аргумент, а значение получаемое из метода.
Если тебе надо проверить что там может вернуться, то надо читать документацию. В этом случае там "возвращает размер". Оно, очевидно, не может быть null или отрицательным. Насчет пустого - если слово не пустое, то маловероятно что размер будет пустой. Но, допустим, ты решил проверить на пустой размер - выкидывать тут исключение с таким типом и текстом не выглядит корректным. Возможно было бы лучше, например, исключить это слово из работы, либо уведомить пользователя что шрифт не имеет символов подходящих (может в этом случае размер будет пустой).

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/TagConfig.cs:
```diff
@@ -0,0 +1,114 @@
+    public Word ConfigureWordTag(Word word, IMainFormConfig mainFormConfig)
+    {
+        word.Size = TextRenderer
+            .MeasureText(word.Value, new Font(_mainFormConfig.FontFamily, word.Weight * _standartSize.Width));
```
`Font` - `IDisposable`.

LevShisterov replied:
Надо диспозить.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/WordConfig.cs:
```diff
@@ -0,0 +1,89 @@
+        if (string.IsNullOrEmpty(word))
+            throw new ArgumentException("Word can not be null or empty");
+        
+        var filePath = Path.Combine(
+            Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.FullName, "boring_words.txt");
```
А тут файл заново для каждого слова загружается? Может можно как-то облегчить?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Forms/TagCloudForm.cs:
```diff
@@ -0,0 +1,51 @@
+    private void SetupWindow()
+    {
+        Text = "Tag Cloud Container";
+        TopMost = true;
```
Это лучше не так делать.
А в месте где вызываешь показ формы:
`tagCloudForm.Show(this);`
Таким образом задается поверх какой формы (окна) надо показать.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/WordConfig.cs:
```diff
@@ -0,0 +1,89 @@
+        return words;
+    }
+
+    private WordConfig For(string word)
```
Можно задать более говорящее имя для метода?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/WordConfig.cs:
```diff
@@ -0,0 +1,89 @@
+            .Where(line => !string.IsNullOrWhiteSpace(line));
+    }
+
+    public List<Word> ShuffleWords(List<Word> words)
```
Есть какое-то ощущение что этот метод тут не к месту.
Хотя, опять же вопрос: а какое назначение у класса `WordConfig`?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/WordConfig.cs:
```diff
@@ -0,0 +1,89 @@
+        return lines
+            .Where(line => !string.IsNullOrWhiteSpace(line))
+            .Select(w => For(w)
+                .RemovePunctuationMarks()
+                .RemoveIfItBoring()
```
В задании есть еще пункт "привести к нижнему регистру".

Luvr681 committed:
Сделал исправления + сборку через autofac

LevShisterov commented on TagCloudContainer/TagCloudContainer/Models/Colors.cs:
```diff
@@ -0,0 +1,28 @@
+    
+    private static readonly Dictionary<string, Color> _colors = new Dictionary<string, Color>()
+    {
+        { "Yellow", (Color)_colorConverter.ConvertFromString("#e9c46a") },
```
А чем `Color.Yellow` не устроило?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Models/Sizes.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    private static readonly Dictionary<string, Size> _sizes = new Dictionary<string, Size>()
+    {
+        { "1920x1080", new Size(1920, 1080) },
```
Тут два набора цифр - а если они разойдутся? Ну, написано для пользователя "1920x1080", а на деле окажется "1980х1020"? Можно как-то избежать?

LevShisterov commented on TagCloudContainer/TagCloudContainer.Tests/TagCloudProvider_Should.cs:
```diff
@@ -0,0 +1,31 @@
+namespace TagCloudContainer.Tests;
+
+[TestFixture]
+public class TagCloudProvider_Should
```
А валидный результат потестим?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Providers/TagCloudProvider.cs:
```diff
@@ -0,0 +1,33 @@
+        var center = MainFormConfig.Center;
+        var standartSize = MainFormConfig.StandartSize;
+        
+        if (center.IsEmpty || center == null)
```
1. Не понимаю зачем мы в этом методе, этом классе, проверяем это значение, если оно в этом классе даже не используется.
2. `center` - это ведь `Point` - `struct` - может быть равно `null`?
3. А почему "пустой" центр не валиден?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/MainFormConfig.cs:
```diff
@@ -0,0 +1,17 @@
+    public static string FontFamily { get; set; } = "Arial";
+    public static Color Color { get; set; } = Colors.Get("Black");
+    public static Color BackgroundColor { get; set; } = Colors.Get("White");
+    public static Size FormSize { get; set; } = Sizes.Get("1920x1080");
```
А если набор размеров поменяют, убрав "1920x1080", а тут забудут?
Можно как-то избежать последствий?

LevShisterov replied:
Я бы посоветовал просто `.First()` - брать первое из доступного. Всё равно пользователь потом из списка будет выбирать.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Forms/TagCloudForm.cs:
```diff
@@ -0,0 +1,53 @@
+        _graphics = e.Graphics;
+        _graphics.Clear(MainFormConfig.BackgroundColor);
+    
+        var pen = new Pen(MainFormConfig.Color);
```
`Pen` - `IDisposable`.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/MainFormConfig.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace TagCloudContainer;
+
+public static class MainFormConfig
```
А почему этот класс статический?
Что если я захочу использовать часть классов и сделать консольного клиента или встроить отрисовщик тегов в какое-либо другое решение, мне придется этот класс так и использовать?

Luvr681 committed:
Внёс правки

LevShisterov commented on TagCloudContainer/TagCloudContainer.Tests/TagCloudProvider_Should.cs:
```diff
@@ -0,0 +1,61 @@
+    [SetUp]
+    public void SetUp()
+    {
+        _mainFormConfig = new MainFormConfig();
```
А зачем тут? Почему не в тестовых методах? `TagCloudProvider` ты ведь там создаешь.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Interfaces/IMainFormConfig.cs:
```diff
@@ -0,0 +1,17 @@
+﻿namespace TagCloudContainer;
+
+public interface IMainFormConfig
```
Хорошо бы разделить этот интерфейс в зависимости от мест использования свойств.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Configs/TagConfig.cs:
```diff
@@ -0,0 +1,110 @@
+        IsValidArguments();
+    }
+    
+    public Word ConfigureWordTag(Word word)
```
Я бы предложил назвать более близко к тому, что делает метод: размещает слово на облаке.
Про `TagConfig` и `TagConfig` - например, в `MainFormConfig` - суффикс говорит что внутри какие-то настройки содержатся.

LevShisterov commented on TagCloudContainer/TagCloudContainer/Program.cs:
```diff
@@ -0,0 +1,14 @@
+
+namespace TagCloudContainer;
+
+public class Program
```
Наверное было бы хорошо вынести формы в отдельный проект, ну, хотя бы чтобы реализовывать консольного клиента было проще.

Luvr681 committed:
Внёс правки, добавил Result<T>

LevShisterov commented on TagCloudContainer/TagCloudContainer.Additions/SizeInvestigator.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class SizeInvestigator : ISizeInvestigator
+{
+    private ITagCloudContainerConfig _tagCloudContainerConfig;
```
`_tagCloudContainerConfig` не используется.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Additions/TagCloudContainer.Additions.csproj:
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <OutputType>WinExe</OutputType>
```
Тут кроется проблема, почему компилятор требует `main`.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Additions/Validator.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        if (tagCloudContainerConfig.Center.IsEmpty)
+            return Result.Fail<ITagCloudContainerConfig>("Center point can't be empty");
+        if (tagCloudContainerConfig.StandartSize.IsEmpty)
+            return Result.Fail<ITagCloudContainerConfig>("Standart font size can't be empty");
```
Проверяется значение `StandartSize`, а в ошибке пишется про "Standart **font** size" - чему верить?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Program.cs:
```diff
@@ -0,0 +1,15 @@
+﻿using Autofac;
```
А этот класс разве не должен быть в одном проекте вместе с собственно формами?

LevShisterov commented on TagCloudContainer/TagCloudContainer/Utils/WordsShuffler.cs:
```diff
@@ -0,0 +1,20 @@
+    public static List<TagCloudContainer.Additions.Models.Word> ShuffleWords(List<TagCloudContainer.Additions.Models.Word> words)
+    {
+        if (words == null)
+            throw new ArgumentException("Argument is null");
```
В исключении не указано название аргумента (для этого есть отдельный параметр).

LevShisterov commented on TagCloudContainer/TagCloudContainer/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,71 @@
+        var filePath = _wordReaderConfig.ExcludeWordsFilePath;
+        
+        _word = new StringBuilder(word.ToLower());
+        _boringWords = File 
+            .ReadLines(filePath)
```
В каждом `ForWord` все еще производится чтение файла: `File.ReadLines`.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/Settings.cs:
```diff
@@ -0,0 +1,91 @@
+
+    private void ValidateUserParameters()
+    {
+        var validator = new Validator();
```
Класс `Validator` не содержит никакого состояния, нет смысла его создавать. Можно было бы сделать статическим, но можно было бы методы переместить в этот (`Settings`) класс.
Но, в `Validator` находятся такие проверки, которые, по идее, должны производиться в конструкторах соответствующих классов.
А сама форма `Settings` маловероятно что произведет кривые настройки, да ей, по сути и всё равно, что там требуют классы, которые эти настройки потребляют.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/Settings.cs:
```diff
@@ -0,0 +1,91 @@
+
+    private void AddConfigValues(bool random)
+    {
+        var parsedUserSelectedSizeValue = Sizes
+            .Text
+            .Split("x")
+            .Select(i => int.Parse(i))
+            .ToArray();
```
Не проще было размеры сразу числами задать, чем потом везде их парсить из строки?

LevShisterov commented on TagCloudContainer/TagCloudContainer.Tests/Register.cs:
```diff
@@ -0,0 +1,25 @@
+
+public static class Register
+{
+    public static IContainer Registry(TagCloudContainerConfig tagCloudContainerConfig, TagCloudFormConfig tagCloudFormConfig, WordReaderConfig wordReaderConfig)
```
Не вижу чтобы это использовалось.

Luvr681 committed:
1. Избавился от throw, сделал через Result
2. Изменил архетиктуру, вынес проект как библиотеку для удобства реализации через cli и ui
3. Вынес конфигурацию в один класс и разделил интерфейсы, которые этот класс реализует

Luvr681 committed:
1. Добавил исключения в конструкторы с проверкой на null
2. Дополнил валидацию конфига
3. Добавил тесты на соответствующую валидацию конфига
4. Написал отлавливание исключений на этапе сборки

LevShisterov commented on TagCloudContainer/TagCloudContainer.Core/Utils/TagCloudPlacer.cs:
```diff
@@ -0,0 +1,123 @@
+            word.Weight * _tagCloudContainerConfig.StandartSize.Width);
+        word.Size = TextRenderer
+            .MeasureText(word.Value, wordFontSize);
+        wordFontSize.Dispose();
```
Используй `using`.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Core/Utils/WordValidator.cs:
```diff
@@ -0,0 +1,73 @@
+        if (tagCloudContainerConfig == null)
+            throw new ArgumentException("Tag cloud config can't be null");
+        
+        _tagCloudContainerConfig = tagCloudContainerConfig;
```
Сейчас модно в виде:
`_tagCloudContainerConfig = tagCloudContainerConfig ?? throw new ArgumentException("Tag cloud config can't be null");`

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/TagCloud.cs:
```diff
@@ -0,0 +1,97 @@
+            font.Dispose();
+        }
+
+        pen.Dispose();
```
`using`
Разница в том, что `using` - это `try {...} finally {.Dispose()}` - т.е. освобождение даже в случае исключения.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/TagCloud.cs:
```diff
@@ -0,0 +1,97 @@
+            Path.Combine(_tagCloudContainerConfig.MainDirectoryPath, _tagCloudContainerConfig.ImageName));
+    }
+
+    public void ValidateConstructorArguments(ITagCloudProvider tagCloudProvider,
+        ITagCloudContainerConfig tagCloudContainerConfig,
+        ITagCloudFormConfig tagCloudFormConfig,
+        IImageCreator imageCreator)
+    {
+        if (tagCloudProvider == null)
+            throw new ArgumentException("Tag cloud provider can't be null");
+        if (tagCloudContainerConfig == null)
+            throw new ArgumentException("Tag cloud config can't be null");
+        if (tagCloudFormConfig == null)
+            throw new ArgumentException("Tag cloud form config can't be null");
+        if (imageCreator == null)
+            throw new ArgumentException("Image creator can't be null");
+    }
```
Лучше в конструкторе
`_tagCloudProvider = tagCloudProvider ?? throw new ArgumentNullException("Tag cloud provider can't be null");`

LevShisterov replied:
**ArgumentNullException**

LevShisterov commented on TagCloudContainer/TagCloudContainer.Core/TagCloudContainer.Core.csproj:
```diff
@@ -0,0 +1,31 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+    <PropertyGroup>
+        <OutputType>WinExe</OutputType>
```
Зачем тут exe?

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/Settings.cs:
```diff
@@ -0,0 +1,118 @@
+        var tagCloudContainerConfigResult = _tagCloudContainerConfigValidator.Validate(_tagCloudContainerConfig);
+        var tagCloudFormConfigResult= _tagCloudFormConfigValidator.Validate(_tagCloudFormConfig);
+        
+        if (!tagCloudContainerConfigResult.IsSuccess)
+            MessageBox.Show("Invalid container options: " + tagCloudContainerConfigResult.Error);
```
Используй `.OnFail()`.

Luvr681 committed:
1. Убрал проверки через if в конструкторе. Заменил на ??;
2. Убрал проверку на Result.IsSuccess. Заменил на Result.OnFail;
3. Внёс dispose в try {} finally {}
4. Размеры экрана теперь в виде Size[]

LevShisterov commented on TagCloudContainer/TagCloudContainer.Forms/Settings.cs:
```diff
@@ -0,0 +1,101 @@
+        IConfigValidator<ITagCloudFormConfig> tagCloudFormConfigValidator,
+        IConfigValidator<ITagCloudContainerConfig> tagCloudContainerConfigValidator)
+    {
+        _tagCloud = tagCloud ?? throw new ArgumentNullException("Tag cloud can't be null");
```
Давай в `ArgumentNullException` еще передадим название параметра.

LevShisterov commented on TagCloudContainer/TagCloudContainer.Core/Utils/TagCloudPlacer.cs:
```diff
@@ -0,0 +1,119 @@
+        }
+        finally
+        {
+            wordFontSize.Dispose();
```
`using var wordFontSize = new Font...`
и [получится автоматический](https://learn.microsoft.com/ru-ru/dotnet/csharp/language-reference/keywords/using-statement) try-finally.

Luvr681 committed:
1. Использовал using Dispose
2. Передал имя переменной в ArgumentException и ArgumentNullException

LevShisterov commented on TagCloudContainer/TagCloudContainer.Core/Utils/TagCloudPlacer.cs:
```diff
@@ -0,0 +1,122 @@
+
+            word.Position = new Point(rectangle.X, rectangle.Y);
+
+            wordFontSize.Dispose();
```
Лишний `.Dispose()`.

Luvr681 committed:
Убрал вызовы Dispose
